### DISCUSSION TITLE
Best way to perform a change of variable within @Integrator framework

### DISCUSSION THREAD

**Discussion by pbouda95** [thelfer/tfel#346] [General]
Hello, 

For a given state variable `foo` which is an unknown of a non linear system, I would like to perform a change of variable inside the integrator code block for its increment `dfoo` only i.e. such as `z=f(dfoo)` for conditioning purposes. However, this is not a global change, so that at the end of each iteration one should be able to retrieve the updaded value of `foo`. Does the `@Implicit` DSL enable to do something like this ?

Cheers,

**Comment by thelfer** [comment]
Hi @pbouda95 

Auxiliary state variables and integration variables are very handy in this case.

You can do the following:

```cxx
....
@AuxiliaryStateVariable real foo;
@IntegrationVariables real z;
....
@InitLocalVariables{
  z = f(foo); 
  ....
}

@Integrator{
  fz = ....
}

@UpdateAuxiliaryStateVariables{
  zoo = inverse_f(z); // z has been updated at this stage !
}
```

See the attached file for details !

[test.zip](https://github.com/thelfer/tfel/files/10057132/test.zip)

**Reply by pbouda95** [reply]
Hi @thelfer ,

I opened your snippet but I am not sure if this will work.
Indeed, in my case I want to perform a change of variable so that `z=dfoo`.
Then, the unknown dfoo is replaced by z and is not an increment, but a variable, and its related increment does not play a role in the NL system.

Did I miss something in your example ?

**Reply by thelfer** [reply]
@pbouda95 In the `ImplicitDSL` you search increments. So your change of variable is `dz=dfoo` with `z` an integration variable. If you don't need the value of `z` (which is usually the value of this variable at the beginning of the time step), don't use it !

At the end of the script in `@UpdateAuxiliaryStateVariables`, just do:

```cxx
  foo += inverse_f(dz);
```

**Reply by pbouda95** [reply]
Wow thanks for that @thelfer  ! 

My troubles were coming from that I was thinking that dz was implicitly defined from the updated value of z and not the opposite ...
Looking back on the script this is much more logical considering the possibility to use the `@UpdateAuxilliaryVariables` block.
I know what I have to do now ...

Cheers,
